lantern_extras/src/lib.rs

```diff
@@ -97,6 +122,73 @@ pub extern "C" fn lantern_daemon_worker() {
     }
 }
 
+#[pg_guard]
+#[no_mangle]
+pub extern "C" fn lantern_indexing_worker() {
+    BackgroundWorker::attach_signal_handlers(SignalWakeFlags::SIGHUP | SignalWakeFlags::SIGTERM);
+    BackgroundWorker::connect_worker_to_spi(Some("postgres"), None);
+
+    let data_dir = BackgroundWorker::transaction(|| {
+        Spi::get_one::<&str>("SELECT setting::text FROM pg_settings WHERE name = 'data_directory'")
+            .expect("Could not query from SPI")
+            .expect("Could not get data data directory")
+    });
+
+    let tmp_dir = std::path::Path::new(data_dir)
+        .join("ldb_indexes")
+        .to_str()
+        .expect("Could not concatinate tmp dir")
+        .to_owned();
+
+    std::fs::create_dir_all(&tmp_dir)
+        .expect(&format!("Could not create tmp directory at {tmp_dir}"));
+
+    let mut started = false;
+
+    while BackgroundWorker::wait_latch(Some(Duration::from_secs(10))) {
+        /* Remove this code after the wait_latch bug is fixed in pgrx */
+        let wakeup_flags = unsafe {
+            let latch = pg_sys::WaitLatch(
+                pg_sys::MyLatch,
+                (pg_sys::WL_TIMEOUT | pg_sys::WL_POSTMASTER_DEATH) as i32,
+                0,
+                pg_sys::PG_WAIT_EXTENSION,
+            );
+            pg_sys::ResetLatch(pg_sys::MyLatch);
+            latch
+        };
+
+        if (wakeup_flags & pg_sys::WL_POSTMASTER_DEATH as i32) != 0 {
+            break;
+        }
+        /* ========================================================== */
+
+        if BackgroundWorker::sighup_received() && !ENABLE_INDEXING_SERVER.get() {
+            std::process::exit(1);
```

Comment:
Why are we exiting with an error code when the external indexing is not enabled?
afaik this causes a postmaster restart, no? is that expected?

Reply:
It is because if the bgworker exits with 0 status it won't be restarted automatically and will be unregistered from postmaster. By exiting the process we make sure that any threads created by this process will be killed and as it exits with non-zero code, it will be restarted after the configured `bgw_restart_time`. After boot it won't start the indexing server (as GUC will be set to false) and will wait for SIGHUP in wait_latch loop, to start in case the `lantern_extras.enable_indexing_server` will be set to true again.

Ref to docs: https://www.postgresql.org/docs/current/bgworker.html#:~:text=If%20bgw_restart_time%20for,itself%20has%20terminated.